### PR TITLE
[lldb/Plugins] Add Swift Runtime Failure Recognizer

### PR DESCRIPTION
**Pull request by medismailben**
This patch add support for Swift Runtime Failure messages in the debugger.

The trap failure messages is wrapped into an artificial inline function
in the debug info, whose name is the actual error message.

When a thread stops in one of these inlined frames, the recognizer fetch
the function name and set it as the stop reason message and select the
parent frame as the current one.

rdar://53003975

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

### CODE REVIEW

**Comment by medismailben** [comment]
@swift-ci Please test OS X platform

File: lldb/source/Plugins/Language/Swift/SwiftRuntimeFailureRecognizer.cpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    SwiftRuntimeFailureRecognizedStackFrame(StackFrameSP most_relevant_frame_sp,
+                                            StringRef stop_desc)
+    : m_most_relevant_frame(most_relevant_frame_sp) {
+  m_stop_desc = stop_desc;
+}
+
+lldb::RecognizedStackFrameSP SwiftRuntimeFailureFrameRecognizer::RecognizeFrame(
+    lldb::StackFrameSP frame_sp) {
+  if (frame_sp->GetFrameIndex())
+    return RecognizedStackFrameSP();
```
Comment (adrian-prantl) [inline lldb/source/Plugins/Language/Swift/SwiftRuntimeFailureRecognizer.cpp]: I usually write this as `return {};`

File: lldb/source/Plugins/Language/Swift/SwiftRuntimeFailureRecognizer.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        "Swift Runtime Failure Recognizer: Hit unwinding bound (1 frame)!");
+    return RecognizedStackFrameSP();
+  }
+
+  SymbolContext sc = frame_sp->GetSymbolContext(eSymbolContextEverything);
+
+  if (!sc.block)
+    return RecognizedStackFrameSP();
+
+  const InlineFunctionInfo *inline_info = nullptr;
```
Comment (adrian-prantl) [inline lldb/source/Plugins/Language/Swift/SwiftRuntimeFailureRecognizer.cpp]: Can you add a comment here explaining that the error is encoded as an inline frame and perhaps even reference the place in the compiler where it is generated so they can be kept in sync?

File: lldb/source/Plugins/Language/Swift/SwiftRuntimeFailureRecognizer.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+void RegisterSwiftRuntimeFailureRecognizer() {
+  static llvm::once_flag g_once_flag;
+
+  // Swift runtime failure
+  llvm::call_once(g_once_flag, []() {
+    RegularExpressionSP module_regex_sp = nullptr;
+    RegularExpressionSP symbol_regex_sp(
+        new RegularExpression("Swift runtime failure"));
+
+    StackFrameRecognizerSP srf_recognizer_sp(
```
Comment (adrian-prantl) [inline lldb/source/Plugins/Language/Swift/SwiftRuntimeFailureRecognizer.cpp]: `std::make_shared<StackFrameRecognizer>()` ?

File: lldb/source/Plugins/Language/Swift/SwiftRuntimeFailureRecognizer.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
+//===-- SwiftRuntimeFailureRecognizer.cpp -----------------------*- C++ -*-===//
```
Comment (adrian-prantl) [inline lldb/source/Plugins/Language/Swift/SwiftRuntimeFailureRecognizer.h]: the `-*- C++ -*-` marker only makes sense for .h file where it is ambiguous whether it's a C or a C++ header.

Comment (adrian-prantl) [reply lldb/source/Plugins/Language/Swift/SwiftRuntimeFailureRecognizer.h]: Oh... it's the filename that's wrong here!

File: lldb/source/Plugins/Language/Swift/SwiftRuntimeFailureRecognizer.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+#ifndef liblldb_SwiftRuntimeFailureRegognizer_h_
+#define liblldb_SwiftRuntimeFailureRegognizer_h_
+
+#include "lldb/Target/StackFrameRecognizer.h"
+
+namespace lldb_private {
+
+void RegisterSwiftRuntimeFailureRecognizer();
+
+/// \class SwiftRuntimeFailureRecognizedStackFrame
```
Comment (adrian-prantl) [inline lldb/source/Plugins/Language/Swift/SwiftRuntimeFailureRecognizer.h]: this line is redundant.

File: lldb/test/Shell/Recognizer/SwiftRuntimeFailure.test [hunk 6]
```diff
@@ -0,0 +1,10 @@
+# UNSUPPORTED: system-windows
+# RUN: %target-swift-frontend -g %S/Inputs/RuntimeFailure.swift -Xllvm -enable-trap-debug-info -o %t.out
+# RUN: %lldb -b -s %s %t.out | FileCheck %s
+run
+# CHECK: thread #{{.*}}stop reason = Swift runtime failure: arithmetic overflow
+frame info
+# CHECK: frame #1{{.*}}`testit{{.*}} at RuntimeFailure.swift
+frame recognizer info 0
+# CHECK: frame 0 is recognized by Swift Runtime Failure StackFrame Recognizer
+q
```
Comment (adrian-prantl) [inline lldb/test/Shell/Recognizer/SwiftRuntimeFailure.test]: I think this should really be a python test, so it's tested with dwarf+dsym and on other platforms

**Review by adrian-prantl — CHANGES_REQUESTED** [review]
Few nitpicks inside, but overall looks good

File: lldb/test/Shell/Recognizer/SwiftRuntimeFailure.test (hunk 6 above)
Comment (adrian-prantl) [reply lldb/test/Shell/Recognizer/SwiftRuntimeFailure.test]: You can start by copying `test/lang/swift/any/`

**Comment by medismailben** [comment]
@swift-ci Please test

File: lldb/packages/Python/lldbsuite/test/lang/swift/runtime_failure_recognizer/TestSwiftRuntimeFailureRecognizer.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        self.expect("process launch", RUN_SUCCEEDED,
+                    substrs=['stop reason = Swift runtime failure: arithmetic overflow'])
+
+        self.expect("frame info",
+                    patterns=['frame #1(.*)`testit(.*)at RuntimeFailure\.swift$'])
+
+        self.expect("frame recognizer info 0",
+                    substrs=['frame 0 is recognized by Swift Runtime Failure StackFrame Recognizer'])
+
+if __name__ == '__main__':
```
Comment (adrian-prantl) [inline lldb/packages/Python/lldbsuite/test/lang/swift/runtime_failure_recognizer/TestSwiftRuntimeFailureRecognizer.py]: You can remove this part. This was useful before we had lldb-dotest, but doesn't really work any more.

File: lldb/source/Plugins/Language/Swift/SwiftRuntimeFailureRecognizer.cpp (hunk 2 above)
Comment (adrian-prantl) [reply lldb/source/Plugins/Language/Swift/SwiftRuntimeFailureRecognizer.cpp]: What I meant was either a filename like "swift/lib/IRGen/IRGenSIL.cpp" or something along the lines of "see the XYZ class in the swift compiler's SILGen library". I don't know how permanent GitHub pull request links are relative to the lifetime of the Swift project.

File: lldb/packages/Python/lldbsuite/test/lang/swift/runtime_failure_recognizer/Makefile [hunk 8]
```diff
@@ -0,0 +1,6 @@
+LEVEL = ../../../make
```
Comment (adrian-prantl) [inline lldb/packages/Python/lldbsuite/test/lang/swift/runtime_failure_recognizer/Makefile]: This should be redundant?

File: lldb/packages/Python/lldbsuite/test/lang/swift/runtime_failure_recognizer/Makefile [hunk 9]
```diff
@@ -0,0 +1,6 @@
...
+
+SWIFT_SOURCES := RuntimeFailure.swift
+SWIFTFLAGS += -Xllvm -enable-trap-debug-info
```
Comment (adrian-prantl) [inline lldb/packages/Python/lldbsuite/test/lang/swift/runtime_failure_recognizer/Makefile]: Is this not on by default?

Comment (medismailben) [reply lldb/packages/Python/lldbsuite/test/lang/swift/runtime_failure_recognizer/Makefile]: Apparently not

File: lldb/packages/Python/lldbsuite/test/lang/swift/runtime_failure_recognizer/Makefile (hunk 8 above)
Comment (medismailben) [reply lldb/packages/Python/lldbsuite/test/lang/swift/runtime_failure_recognizer/Makefile]: I tried to remove it, with the include (few lines below) but I keep getting Unexpected Failures.

**Comment by medismailben** [comment]
@swift-ci please test

**Comment by medismailben** [comment]
@swift-ci please test

**Review by adrian-prantl — APPROVED** [review]

**Comment by medismailben** [comment]
@swift-ci please test

**Comment by fredriss** [comment]
This needs to go to `swift/master` `master-rebranch` has been merged yesterday

**Comment by medismailben** [comment]
The automerger still merges `apple/stable/20200108` to `swift/master-rebranch`.
Since I landed #765 on `apple/stable/20200108`, I want to make sure the bot failures are fixed.

**Pull request closed without merging**